### PR TITLE
Optimize rendering of ThreadStateBar

### PR DESCRIPTION
**Pull request by karupayun**
In this PR we are following the plan designed in go/stadia-orbit-rendering-60fps for ThreadStates. We are optimizing its rendering by iterating through pixels as we recently implemented for both SchedulerTrack (https://github.com/google/orbit/pull/4156) and CallstackThreadBar (https://github.com/google/orbit/pull/4271).

The implementation is quite similar to CallstackThreadBar:
- ForEach...Discretized method in CaptureData who iterates through pixels using lower_bound queries.
- UnitTests in CaptureDataTest.
- A completely simplified implementation in ThreadStateBar that use the new method instead of having its own implementation

As a result we are rendering each ThreadStateBar up to 4x times faster (1.2ms instead of 4.8ms for some with several events) when zooming-out in the 5-min capture introduced in the design doc.

Overall rendering time (5min capture - 1200 pixels width - fully zoomed out, fully scaled so every Track is rendered):
Before: ~175.999ms (http://screenshot/crXutG4s66CSETh)
After: ~118.819ms (http://screenshot/9YyvGUmkk8Paeuv)
Improvement: ~67.5%

The new visual appearence can be seen here: http://screenshot/6iLSpkSr5QJc3Ka
We are also fixing an issue. Before this PR blue slices has bigger priority to be rendered and were more visible misleading the user (Border case: http://screenshot/5CNdZd89H5wn9pi). Now we are taking the first one at each pixel. This isn't perfect (because anti-aliasing), but it's still better than the current implementation.

Bug: http://b/253429871
Test: Load a capture. Zoom-out entirely, zoom-in to the maximum. See ThreadStates rendering looks as it's expected.

### CODE REVIEW

File: src/ClientData/CaptureDataTest.cpp [hunk 1]
```diff
@@ -55,10 +58,15 @@ constexpr uint64_t kSecondTid = 2000;
 constexpr uint64_t kNonExistingTid = 404;
 constexpr uint64_t kStTimestamp1 = 50;
 constexpr uint64_t kEnTimestamp1 = 100;
```
Comment (dpallotti) [inline src/ClientData/CaptureDataTest.cpp:60]: Not part of this change, but do "st" and "en" mean "start" and "end"? Avoid custom abbreviations as per the style guide.

File: src/ClientData/CaptureDataTest.cpp [hunk 2]
```diff
@@ -271,6 +279,63 @@ TEST_F(CaptureDataTest, UpdateTimerDurationsIsCorrect) {
   EXPECT_THAT(capture_data_.GetSortedTimerDurationsForScopeId(kNotIssuedId), testing::IsNull());
 }
 
+struct FormattingForEachThreadStateSliceIntersectingTimeRangeDiscretizedTest {
```
Comment (dpallotti) [inline src/ClientData/CaptureDataTest.cpp:282]: What does "formatting" mean here? I feel like it's the wrong word.

What about `ForEachThreadStateSliceIntersectingTimeRangeDiscretizedTestCase`?

File: src/ClientData/include/ClientData/CaptureData.h [hunk 3]
```diff
@@ -127,6 +127,11 @@ class CaptureData {
       uint32_t thread_id, uint64_t min_timestamp, uint64_t max_timestamp,
       const std::function<void(const ThreadStateSliceInfo&)>& action) const;
 
+  // Similar to the previous one, but does not iterate for more than one slice per pixel.
```
Comment (dpallotti) [inline src/ClientData/include/ClientData/CaptureData.h:130]:
```suggestion
  // Similar to the previous one, but does not iterate on more than one slice per pixel.
```
or "over"
?

**Review by dpallotti — APPROVED** [review]

**Review by florian-kuebler — APPROVED** [review]
LGTM. Thanks!

File: src/ClientData/CaptureDataTest.cpp (hunk 2 above)
Comment (karupayun) [reply src/ClientData/CaptureDataTest.cpp:282]: I took it from this example:
https://engdoc.corp.google.com/eng/doc/devguide/testing/parameterized-testing.md?cl=head&polyglot=cpp

But I can change it, I'm also not sure that I like this "formatting".

Anyway, I've already used in the CallstackData new tests so I may need to change it there as well, so it will be part of a following PR.

Comment (dpallotti) [reply src/ClientData/CaptureDataTest.cpp:282]: It's called "Formatting" there because the tests are testing a class called `Formatter`...
